lib/runtime/storage/trie.go

```diff
@@ -17,97 +17,129 @@ import (
 // TrieState is a wrapper around a transient trie that is used during the course of executing some runtime call.
 // If the execution of the call is successful, the trie will be saved in the StorageState.
 type TrieState struct {
-	t       *trie.Trie
-	oldTrie *trie.Trie // this is the trie before BeginStorageTransaction is called. set to nil if it isn't called
-	lock    sync.RWMutex
+	mtx          sync.RWMutex
+	transactions []*trie.Trie
+	currentTrie  *trie.Trie
 }
 
-// NewTrieState returns a new TrieState with the given trie
-func NewTrieState(t *trie.Trie) *TrieState {
-	if t == nil {
-		t = trie.NewEmptyTrie()
-	}
-
+func NewTrieState(state *trie.Trie) *TrieState {
 	return &TrieState{
-		t: t,
+		transactions: make([]*trie.Trie, 0),
+		currentTrie:  state,
 	}
 }
 
-// Trie returns the TrieState's underlying trie
-func (s *TrieState) Trie() *trie.Trie {
-	return s.t
+func (t *TrieState) getCurrentTrie() *trie.Trie {
```

Comment:
We have a mixed behavior here, if we always hold 1 transaction we can be sure that the last item is the currentTrie and this method will be trivial